src/sentry/scripts/digests/digests.lua

```diff
@@ -0,0 +1,318 @@
+        -- should always exist, and we established that above) but not
+        -- using that here doesn't break anything and allows use to use
+        -- older Redis versions.
+        redis.call('ZADD', configuration:get_schedule_waiting_key(), update, timeline_id)
+    end
+
+    return false
+end
+
+local function truncate_timeline(configuration, timeline_id, capacity)
+    return zset_trim(
+        configuration:get_timeline_key(timeline_id),
+        capacity,
+        function (record_id)
+            redis.call('DEL', configuration:get_timeline_record_key(timeline_id, record_id))
+        end
+    )
+end
+
+local function truncate_digest(configuration, timeline_id, capacity)
+    return zset_trim(
+        configuration:get_timeline_digest_key(timeline_id),
+        capacity,
+        function (record_id)
+            redis.call('DEL', configuration:get_timeline_record_key(timeline_id, record_id))
+        end
+    )
+end
+
+local function add_record_to_timeline(configuration, timeline_id, record_id, value, timestamp, delay_increment, delay_maximum, timeline_capacity, truncation_chance)
+    redis.call('SETEX', configuration:get_timeline_record_key(timeline_id, record_id), configuration.ttl, value)
+    redis.call('ZADD', configuration:get_timeline_key(timeline_id), timestamp, record_id)
+    redis.call('EXPIRE', configuration:get_timeline_key(timeline_id), configuration.ttl)
+
+    local ready = add_timeline_to_schedule(configuration, timeline_id, timestamp, delay_increment, delay_maximum)
+
+    -- TODO: `tonumber` should happen upstream
+    local timeline_capacity = tonumber(timeline_capacity)
+    if timeline_capacity > 0 and math.random() < tonumber(truncation_chance) then
+        -- TODO: `tonumber` should happen upstream
+        truncate_timeline(configuration, timeline_id, timeline_capacity)
+    end
+
+    return ready
+end
+
+local function digest_timeline(configuration, timeline_id)
+    -- Check to ensure that the timeline is in the correct state.
+    if redis.call('ZSCORE', configuration:get_schedule_ready_key(), timeline_id) == false then
+        error('err(invalid_state): timeline is not in the ready state, cannot be digested')
+    end
+
+    local digest_key = configuration:get_timeline_digest_key(timeline_id)
+    local timeline_key = configuration:get_timeline_key(timeline_id)
+    if redis.call('EXISTS', digest_key) == 1 then
+        -- If the digest set already exists (possibly because we already tried
+        -- to send it and failed for some reason), merge any new data into it.
+        -- TODO: It might make sense to trim here to avoid returning capacity *
+        -- 2 if timeline was full when it was previously digested.
+        redis.call('ZUNIONSTORE', digest_key, 2, timeline_key, digest_key, 'AGGREGATE', 'MAX')
+        redis.call('DEL', timeline_key)
+        redis.call('EXPIRE', digest_key, configuration.ttl)
+    elseif redis.call('EXISTS', timeline_key) == 1 then
```

Comment:
In the previous block we already assumed the `timeline_key` must exist, since we're using `ZUNIONSTORE` to merge the timeline key in. Why check if it exists in this conditional? What if it doesn't exist, we don't have created the digest key right?